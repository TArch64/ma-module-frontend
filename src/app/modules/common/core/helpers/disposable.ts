import { Observable, OperatorFunction, Subject, PartialObserver } from 'rxjs';
import { takeUntil } from 'rxjs/operators';

export type NextOrObserver<T> = PartialObserver<T> | ((value: T) => void);

export class Disposable {
    private subscriptionDisposingTrigger = new Subject<null>();

    public takeUntilDisposed<T>(): OperatorFunction<T, T> {
        return takeUntil<T>(this.subscriptionDisposingTrigger);
    }

    public subscribeTo<T>(observable: Observable<T>, nextOrObserver: NextOrObserver<T>): void {
        const safeObservable = observable.pipe(this.takeUntilDisposed());

        if (typeof nextOrObserver === 'function') {
            safeObservable.subscribe(nextOrObserver);
            return;
        }
        safeObservable.subscribe(nextOrObserver);
    }

    public dispose(): void {
        this.subscriptionDisposingTrigger.next(null);
        this.subscriptionDisposingTrigger.complete();
    }
}
