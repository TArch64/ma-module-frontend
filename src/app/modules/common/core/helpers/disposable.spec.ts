import {Subject} from "rxjs";
import {toArray} from "rxjs/operators";
import { Disposable } from './disposable';

describe('take until disposed', () => {
    test('should take events until disposed', (done) => {
        const disposable = new Disposable();
        const subject = new Subject();

        subject.pipe(disposable.takeUntilDisposed(), toArray()).subscribe({
            next: values => expect(values).toMatchSnapshot(),
            complete: () => done()
        });

        subject.next('first');
        subject.next('second');
        disposable.dispose();
        subject.next('ignored');
    });
});

describe('subscribe to', () => {
    test('should subscribe until disposed', (done) => {
        const disposable = new Disposable();
        const subject = new Subject<string>();
        const values: string[] = [];

        disposable.subscribeTo(subject, {
            next: value => values.push(value),

            complete() {
                expect(values).toMatchSnapshot();
                done();
            }
        });

        subject.next('first');
        subject.next('second');
        disposable.dispose();
        subject.next('ignored');
    });

    test('should subscribe until disposed via next callback', (done) => {
        const disposable = new Disposable();
        const subject = new Subject<string>();
        const values: string[] = [];

        disposable.subscribeTo(subject, value => values.push(value));

        disposable.subscribeTo(subject, {
            complete() {
                expect(values).toMatchSnapshot();
                done();
            }
        })

        subject.next('first');
        subject.next('second');
        disposable.dispose();
        subject.next('ignored');
    });
});
