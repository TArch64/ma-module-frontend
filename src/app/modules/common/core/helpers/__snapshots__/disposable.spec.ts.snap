// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscribe to should subscribe until disposed 1`] = `
Array [
  "first",
  "second",
]
`;

exports[`subscribe to should subscribe until disposed via next callback 1`] = `
Array [
  "first",
  "second",
]
`;

exports[`take until disposed should take events until disposed 1`] = `
Array [
  "first",
  "second",
]
`;
